libraries/AP_HAL_ChibiOS/Semaphores.h

```diff
@@ -37,3 +37,17 @@ class ChibiOS::Semaphore : public AP_HAL::Semaphore {
     // we declare the lock as a uint32_t array, and cast inside the cpp file
     uint32_t _lock[5];
 };
+
+class ChibiOS::CountingSemaphore : public AP_HAL::CountingSemaphore {
+public:
+    CountingSemaphore(uint8_t initial_count=0);
+
+    bool wait(uint32_t timeout_us) override;
+    bool wait_blocking(void) override;
+    void signal(void) override;
+    void signal_ISR(void) override;
+    uint8_t get_count(void) override;
+
+protected:
+    uint32_t _lock[5];
```

Comment:
why not include "ch.h" and use semaphore_t and mutex_t. I double checked it seems to be compiling fine with that. Maybe it was done to go around a circular include, it doesn't seem to be the case anymore.

Reply:
for the same reason we use this mechanism for the mutex - it makes building faster. Including ch.h costs quite a lot in a header that ends up being included everywhere. We originally used ch.h in Semaphores.h then moved away from it to speed up builds